py4DSTEM/io/datastructure/datacube.py

```diff
@@ -286,7 +287,15 @@ def get_datacube_from_grp(g,mem='RAM',binfactor=1,bindtype=None):
     """ Accepts an h5py Group corresponding to a single datacube in an open, correctly formatted H5 file,
         and returns a DataCube.
     """
-    # TODO: add memmapping, binning
+    # TODO: add binning
+    if mem == 'MEMMAP':
+        h5_filename = g.file.filename
+        with emd.fileEMD(h5_filename) as emdF:
+            emd_memmap = emdF.get_memmap(0)
+            dataset = emd_memmap[0]
+            mm = np.memmap(h5_filename, shape=dataset.shape, dtype=dataset.dtype, offset=dataset.id.get_offset())
```

Comment:
Brilliant approach here: https://groups.google.com/g/h5py/c/1Fjt19OPcYY/m/HoXXJARwAgAJ

Reply:
The memmap returned by ncempy's function `get_memmap()` uses h5py for accessing the dataset directly from disk. You are providing instead a numpy memmap to the binary chunk on disk. This requires that the data is contiguous on disk (i.e. the h5py dataset does not use chunking or filtering in h5py) to work properly.

It might be better to just allow h5py to provide the memmap instead, because then the user does not have to worry about data contiguity. h5py takes care of that for you. So, you can just return `emd_memmap[0]`.

Reply:
Thanks for the feedback! That makes sense!
It looks like making this change poses challenges. The current expectation, as I understand it, is:
- Passing around a DataCube instance
- DataCube.data is an ndarray type

Given that the entrypoint for this logic is `py4dstem.io.read()`, returning `h5py.Dataset` for MEMMAP cases would break backward compatibility for anyone who chooses to pass in MEMMAP. 

- Does it make sense to change the purpose of this pull request so it's clear that this change would only support contiguous datasets for now? I could add some assertions to gate usage accordingly and continue thinking about a more complete solution for another pull request.
- Are there other approaches I could take that I'm missing? For e.g., maybe adding the logic here isn't right, but somewhere else might work?
  - Something else I noticed is that when I pass around the get_memmap() result, the DataSet is closed after returning the result from the function. Looks like it got garbage collected and closed accordingly.

I committed my latest attempt for visibility. I'm still learning my way around, so I appreciate the feedback!

Reply:
> Given that the entrypoint for this logic is py4dstem.io.read(), returning h5py.Dataset for MEMMAP cases would break backward compatibility for anyone who chooses to pass in MEMMAP.

Im not sure this is true. Most things that are done in numpy can be done on a hdf5_dataset object equally. Not every function is allowed (i.e. reshape). So, it still might not break anything but is only a partial "drop-in" replacement.

> Does it make sense to change the purpose of this pull request so it's clear that this change would only support contiguous datasets for now? I could add some assertions to gate usage accordingly and continue thinking about a more complete solution for another pull request.

In my opinion: yes. But its up to maintainers of py4DSTEM to weigh in on how they want to support HDF5 memmaps. Im just providing context from the `ncempy` reading side which is being used under the hood in py4DSTEM.

> Something else I noticed is that when I pass around the get_memmap() result, the DataSet is closed after returning the result from the function. Looks like it got garbage collected and closed accordingly.

This is because you are using the `with` context manager. This is suggested to avoid accidentally leaving a file open in other cases. Not in this instance, however. Here you are opening the file, returning the hdf5_dataset object, and finally closing the file. For any memmap (hdf5, numpy, etc.) you have to have the file open to keep accessing it on disc. Remove the `with` to solve that problem.